File: homework-1/backend-fastapi/app/ratelimit.py
import threading
import time

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse


class RateLimitMiddleware(BaseHTTPMiddleware):
    """In-memory, per-IP fixed-window rate limiter (Task 4D)."""

    def __init__(self, app, limit: int = 100, window_seconds: int = 60) -> None:
        super().__init__(app)
        self.limit = limit
        self.window = window_seconds
        self._hits: dict[str, list] = {}
        self._lock = threading.Lock()

    async def dispatch(self, request, call_next):
        ip = request.client.host if request.client else "unknown"
        now = time.time()
        with self._lock:
            window = self._hits.get(ip)
            if window is None or now - window[0] >= self.window:
                self._hits[ip] = [now, 1]
                allowed = True
            else:
                window[1] += 1
                allowed = window[1] <= self.limit

        if not allowed:
            return JSONResponse(
                status_code=429,
                content={
                    "error": "Too Many Requests",
                    "message": f"Rate limit of {self.limit} requests per minute exceeded",
                },
            )
        return await call_next(request)

File: homework-1/backend-go/internal/httpapi/ratelimit.go
package httpapi

import (
	"fmt"
	"net"
	"net/http"
	"sync"
	"time"
)

type hitWindow struct {
	start time.Time
	count int
}

// rateLimiter is an in-memory, per-key fixed-window limiter (Task 4D).
type rateLimiter struct {
	mu     sync.Mutex
	limit  int
	window time.Duration
	hits   map[string]*hitWindow
}

func newRateLimiter(limit int, window time.Duration) *rateLimiter {
	return &rateLimiter{limit: limit, window: window, hits: make(map[string]*hitWindow)}
}

func (rl *rateLimiter) allow(key string) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	now := time.Now()
	h := rl.hits[key]
	if h == nil || now.Sub(h.start) >= rl.window {
		rl.hits[key] = &hitWindow{start: now, count: 1}
		return true
	}
	h.count++
	return h.count <= rl.limit
}

// rateLimitMiddleware limits each client IP to `limit` requests per `window`.
func rateLimitMiddleware(limit int, window time.Duration) func(http.Handler) http.Handler {
	rl := newRateLimiter(limit, window)
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !rl.allow(clientIP(r)) {
				writeJSON(w, http.StatusTooManyRequests, map[string]any{
					"error":   "Too Many Requests",
					"message": fmt.Sprintf("Rate limit of %d requests per minute exceeded", limit),
				})
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func clientIP(r *http.Request) string {
	if host, _, err := net.SplitHostPort(r.RemoteAddr); err == nil {
		return host
	}
	return r.RemoteAddr
}
